### PR TITLE
WIP refactor pending dependencies

### PR DESCRIPTION
**Pull request by bigdaz**

### CODE REVIEW

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/DefaultArtifactDependencyResolver.java [hunk 1]
```diff
@@ -95,14 +94,14 @@ public void resolve(ResolveContext resolveContext, List<? extends ResolutionAwar
...
-            new CachingDependencySubstitutionApplicator(new DefaultDependencySubstitutionApplicator(resolutionStrategy.getDependencySubstitutionRule()));
-        DependencyToComponentIdResolver componentIdResolver = new DependencySubstitutionResolver(componentSource.getComponentIdResolver(), applicator);
+        DependencyToComponentIdResolver componentIdResolver = componentSource.getComponentIdResolver();
         ComponentMetaDataResolver componentMetaDataResolver = new ClientModuleResolver(componentSource.getComponentResolver(), dependencyDescriptorFactory);
 
         ResolveContextToComponentResolver requestResolver = createResolveContextConverter();
         ConflictHandler conflictHandler = createConflictHandler(resolutionStrategy, globalRules);
 
+        DependencySubstitutionApplicator applicator =
+            new CachingDependencySubstitutionApplicator(new DefaultDependencySubstitutionApplicator(resolutionStrategy.getDependencySubstitutionRule()));
```
Comment (melix) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/DefaultArtifactDependencyResolver.java]: :o: This is something to check, but I _think_ that with your changes, now, we can get rid of the `CachingDependencySubstitutionApplicator`. Which should remove a bit of overhead in dependency management and make things faster. If it's not the case, if I remember well, at least one test is going to fail, checking that a rule is only applied once.

Comment (bigdaz) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/DefaultArtifactDependencyResolver.java]: Thanks: I had the same thought, but removing the cache causes multiple application of rules in many of our tests, which is not something we want. I might look back and see how we were avoiding multiple rule application _before_ the `CachingDependencySubstitutionApplicator` was introduced.

**Comment by jjohannes** [comment]
We rebaselined the performance tests on master. I would suggest to rebase this PR. Run the performance tests again and have a close look at the results to check if these changes here really cause an additional regression.

**Comment by bigdaz** [comment]
Thanks @jjohannes, the PR merged smoothly the second time around.

**Pull request closed without merging**